rir/src/compiler/analysis/generic_static_analysis.h

```diff
@@ -336,7 +336,320 @@ class StaticAnalysis {
+        snapshots.resize(code->nextBBId);
+        for (auto e : cfg.exits()) {
+            entrypoints.push_back(e);
+        }
+    }
+    BackwardStaticAnalysis(const std::string& name, ClosureVersion* cls,
+                           Code* code, const AbstractState& initialState,
+                           CFG const& cfg, LogStream& log)
+        : name(name), cfg(cfg), log(log), closure(cls), code(code) {
+        snapshots.resize(code->nextBBId);
+        for (auto e : cfg.exits()) {
+            entrypoints.push_back(e);
+            snapshots[e->id].entry = initialState;
+        }
+    }
+
+    const AbstractState& result() const {
+        assert(done);
+        return exitpoint;
+    }
+
+    void logHeader() const {
+        if (DEBUG_LEVEL > AnalysisDebugLevel::None) {
+            log << "=========== Starting " << name << " Analysis on "
+                << closure->name() << " ";
+            if (code == closure)
+                log << "body";
+            else
+                log << "Prom(" << static_cast<Promise*>(code)->id << ")";
+            log << "\n";
+        }
+    }
+
+    void logInitialState(const AbstractState& state, const BB* bb) const {
+        if (DEBUG_LEVEL >= AnalysisDebugLevel::BB) {
+            log << "======= Entering BB" << bb->id << ", initial state\n";
+            log(state);
+        }
+    }
+
+    void logChange(const AbstractState& post, const AbstractResult& res,
+                   const Instruction* i) {
+        if (DEBUG_LEVEL >= AnalysisDebugLevel::Instruction &&
+            res >= AbstractResult::None) {
+            log << "===== After applying instruction ";
+            if (res == AbstractResult::Tainted) {
+                log << " (State got tainted)";
+            } else {
+                log(i);
+            }
+            log << " we have\n";
+            log(post);
+        }
+    }
+
+    void logTaintChange(const AbstractState& pre, const AbstractState& post,
+                        const AbstractResult& res, const Instruction* i) {
+        assert(DEBUG_LEVEL == AnalysisDebugLevel::Taint);
+        if (res >= AbstractResult::Tainted) {
+            if (res == AbstractResult::Tainted) {
+                log << "===== Before applying instruction ";
+                log(i);
+                log << " we have\n";
+                log(pre);
+            }
+            log << "===== After applying instruction ";
+            if (res == AbstractResult::Tainted) {
+                log << " (State got tainted)";
+            } else {
+                log(i);
+            }
+            log << " we have\n";
+            log(post);
+        }
+    }
+
+    void logExit(const AbstractState& state) {
+        if (DEBUG_LEVEL >= AnalysisDebugLevel::Exit) {
+            log << "===== Exit state is\n";
+            log(state);
+        }
+    }
+
+    template <PositioningStyle POS>
+    AbstractState at(Instruction* i) const {
```

Comment:
are you using this api? if not, remove, because it is out of sync with the forward version.

Reply:
It's dumb to keep the two separate in the first place, it was supposed to be a quick temporary fix but it stuck and now is the second time I have to update :)

I guess they would be better unified, similar to the visitors..

Reply:
Couldn't find a nice way to unify them... the backward is mostly copy paste of the forward :(
And what's more, I now use the liveness that we already had, so the backward analysis is not used anywhere anymore.. Remove?